src/main/java/com/mraof/minestuck/alchemy/GristHelper.java

```diff
@@ -135,58 +141,61 @@ public static void increase(Level level, PlayerIdentifier player, GristSet set)
 		data.setGristCache(newCache);
 	}
 	
-	public static void notify(MinecraftServer server, PlayerIdentifier player, GristSet set)
+	public static void notify(MinecraftServer server, PlayerIdentifier player, GristSet set, GristToast.EnumSource source, boolean increase)
 	{
 		if(MinestuckConfig.SERVER.showGristChanges.get())
 		{
-			Map<GristType, Long> reqs = set.getMap();
-			for(Entry<GristType, Long> pairs : reqs.entrySet())
+			
+			//don't know why this is here, it was copied over from when i merged notifyEditMode() and notify(), but im not sure what it does/did. Seems to be a sanity check.
+			SburbConnection sc;
+			EditData ed;
+			
+			if(source == GristToast.EnumSource.SERVER) {
+				sc = SkaianetHandler.get(server).getActiveConnection(player);
+				if(sc == null)
+					return;
+				
+				ed = ServerEditHandler.getData(server, sc);
+				if(ed == null)
+					return;
+			}
+			
+			if(player.getPlayer(server).isLocalPlayer() == true)
 			{
-				Component type = pairs.getKey().getDisplayName();
-				long difference = pairs.getValue();
-				sendGristMessage(server, player, new TranslatableComponent("You gained %s %s grist.", difference, type));
+				sendGristMessage(player.getPlayer(server), set, source, increase);
+			}
+			else
+			{
+				GristToastPacket gristToastPacket = new GristToastPacket(set, source, increase);
+				MSPacketHandler.sendToPlayer(gristToastPacket, player.getPlayer(server));
 			}
+			
+			//sendGristMessage(server, player, new TranslatableComponent("You gained %s %s grist.", difference, type));
 		}
 	}
 	
-	public static void notifyEditPlayer(MinecraftServer server, PlayerIdentifier player, GristSet set, boolean increase)
+	@OnlyIn(Dist.CLIENT)
+	public static void sendGristMessage(Player player, GristSet set, GristToast.EnumSource source, boolean increase)
```

Comment:
This function is more appropriate for `GristToast`.

Reply:
I put it in GristHelper because GristSet.getMap() is protected for some reason, and I think it's easier to read this way.

Reply:
You could use `gristSet.getAmounts()` instead of the map.

Reply:
Alright, it's been moved to GristToast.